primitives/src/merkle_tree/namespaced_merkle_tree.rs

```diff
@@ -0,0 +1,262 @@
+pub struct NamespacedHasher<H, E, I, T, N>
+where
+    H: DigestAlgorithm<E, I, T>,
+    E: Element + Namespaced<N>,
+    N: Namespace,
+    I: Index,
+    T: NodeValue,
+{
+    phantom1: PhantomData<H>,
+    phantom2: PhantomData<E>,
+    phantom3: PhantomData<I>,
+    phantom4: PhantomData<T>,
+    phantom5: PhantomData<N>,
+}
+
+/// Trait indicating that a leaf has a namespace
+pub trait Namespaced<N: Namespace> {
+    /// Returns the namespace of the leaf
+    fn get_namespace(&self) -> N;
+}
+
+/// Trait indicating that a digest algorithm can bind namespaces
+pub trait BindNamespace<E, I, T, N>: DigestAlgorithm<E, I, T>
+where
+    E: Element,
+    N: Namespace,
+    T: NodeValue,
+    I: Index,
+{
+    /// Generate a commitment that binds a node to a namespace range
+    fn generate_namespaced_commitment(namespaced_hash: NamespacedHash<T, N>) -> T;
+}
+
+/// Trait indiciating that a struct can act as an orderable namespace
+pub trait Namespace:
+    Debug + Clone + CanonicalDeserialize + CanonicalSerialize + Default + Copy + Hash + Ord
+{
+    /// Returns the minimum possible namespace
+    fn min() -> Self;
+    /// Returns the maximum possible namespace
+    fn max() -> Self;
+}
+
+impl Namespace for u64 {
+    fn min() -> u64 {
+        u64::MIN
+    }
+    fn max() -> u64 {
+        u64::MAX
+    }
+}
+
+#[derive(
+    CanonicalSerialize,
+    CanonicalDeserialize,
+    Hash,
+    Copy,
+    Clone,
+    Debug,
+    Ord,
+    Eq,
+    PartialEq,
+    PartialOrd,
+)]
+/// Represents a namespaced internal tree node
+pub struct NamespacedHash<T, N>
+where
+    N: Namespace,
+    T: NodeValue,
+{
+    min_namespace: N,
+    max_namespace: N,
+    hash: T,
+}
+
+impl<T, N> Default for NamespacedHash<T, N>
+where
+    N: Namespace,
+    T: NodeValue,
+{
+    fn default() -> Self {
+        Self {
+            hash: T::default(),
+            max_namespace: <N as Namespace>::min(),
+            min_namespace: <N as Namespace>::max(),
+        }
+    }
+}
+
+impl<T, N> NamespacedHash<T, N>
+where
+    N: Namespace,
+    T: NodeValue,
+{
+    /// Constructs a new NamespacedHash
+    pub fn new(min_namespace: N, max_namespace: N, hash: T) -> Self {
+        Self {
+            min_namespace,
+            max_namespace,
+            hash,
+        }
+    }
+}
+
+impl<E, H, T, I, N> DigestAlgorithm<E, I, NamespacedHash<T, N>> for NamespacedHasher<H, E, I, T, N>
+where
+    E: Element + Namespaced<N>,
+    I: Index,
+    N: Namespace,
+    T: NodeValue,
+    H: DigestAlgorithm<E, I, T> + BindNamespace<E, I, T, N>,
+{
+    fn digest(data: &[NamespacedHash<T, N>]) -> NamespacedHash<T, N> {
+        if data.is_empty() {
+            return NamespacedHash::default();
+        }
+        let first_node = data[0];
+        let min_namespace = first_node.min_namespace;
+        let mut max_namespace = first_node.max_namespace;
+        let mut nodes = vec![H::generate_namespaced_commitment(first_node)];
+        for node in data {
```

Comment:
Should not we skip the first element in the loop? That is:  `for node in data.skip(1)` ?